File: Easy/Merge_Two_Sorted_Lists.cpp
/*
Prompt:
Merge two sorted linked lists and return it as a sorted list. The list should be made by splicing together the nodes of the first two lists.

*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if(l1 == NULL){
            return l2;
        }
        if(l2 == NULL){
            return l1;
        }
        
        ListNode *root;
        if(l1->val <= l2->val){
            root = l1;
            root->next = mergeTwoLists(l1->next, l2);
        }
        else{
            root = l2;
            root->next = mergeTwoLists(l1, l2->next);
        }
        return root;
    }
};

/*
Notes:
This way uses recursion and is very fast. Another way to do this is to loop while boths lists have nodes
and in each loop iteration compare the current of one list with the current val of the other list

Constraints:
The number of nodes in both lists is in the range [0, 50].
-100 <= Node.val <= 100
Both l1 and l2 are sorted in non-decreasing order.

*/

File: Easy/Merge_Two_Sorted_Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        newList = ListNode()
        newListHead = newList
        shorterList = 0


        if not list1:
            return list2
        if not list2:
            return list1

        if list1.val >= list2.val:
            newList.val = list2.val
            list2 = list2.next
            shorterList = 2
            if not list2:
                newListHead.next = list1
                return newList
        else:
            newList.val = list1.val
            list1 = list1.next
            shorterList = 1
            if not list1:
                newListHead.next = list2
                return newList
            

        while list1 and list2:
            tempList = ListNode()
            if list1.val >= list2.val:
                tempList = list2
                newListHead.next = tempList
                newListHead = newListHead.next
                list2 = list2.next
                if not list2:
                    shorterList = 2
                    break
            else:
                tempList.val = list1.val
                newListHead.next = tempList
                newListHead = newListHead.next
                list1 = list1.next
                if not list1:
                    shorterList = 1
                    break
                
            
        if shorterList == 1:
            newListHead.next = list2
        else:
            newListHead.next = list1
           

        return newList
        